cli: add more choices for formatting table results.

knz opened the pull request:
This patch replaces the boolean flag `--pretty` by a new string-valued
flag `--format`.

Possible values: `tsv`, `pretty`, `csv`, `html`, `records` (PostgreSQL-style), `sql`.

As previously, `--format=tsv` is the default for non-interactive
sessions, and `--format=pretty` is the default for interactive
sessions.

Fixes #12985.

Examples:

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=tsv
```
```
1 row
a	b
3       4
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=pretty
```
```
+---+---+
| a | b |
+---+---+
| 3 | 4 |
+---+---+
(1 row)
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=csv
```
```csv
1 row
a,b
3,4
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=html
```
```html
<table>
<thead><tr><th>a</th><th>b</th></tr></head>
<tbody>
<tr><td>3</td><td>4</td></tr>
</tbody>
</table>
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=records
```
```
-[ RECORD 1 ]
a | 3
b | 4
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=sql
```
```sql
CREATE TABLE results (
  a STRING,
  b STRING
);

INSERT INTO results VALUES ('3', '4');
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13171)
<!-- Reviewable:end -->

knz committed:
cli: add more choices for formatting table results.

This patch replaces the boolean flag `--pretty` by a new string-valued
flag `--format`.

Possible values: `tsv`, `pretty`, `csv`, `html`, `records`
(similar to PostgreSQL's "expanded" format), `sql`.

As previously, `--format=tsv` is the default for non-interactive
sessions, and `--format=pretty` is the default for interactive
sessions.

Examples:

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=tsv
```
```
1 row
a	b
3       4
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=pretty
```
```
+---+---+
| a | b |
+---+---+
| 3 | 4 |
+---+---+
(1 row)
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=csv
```
```csv
1 row
a,b
3,4
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=html
```
```html
<table>
<thead><tr><th>a</th><th>b</th></tr></head>
<tbody>
<tr><td>3</td><td>4</td></tr>
</tbody>
</table>
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=records
```
```
-[ RECORD 1 ]
a | 3
b | 4
```

```shell
./cockroach sql -e "SELECT 3 AS a, 4 AS b" --format=sql
```
```sql
CREATE TABLE results (
  a STRING,
  b STRING
);

INSERT INTO results VALUES ('3', '4');
```

bdarnell commented on pkg/cli/cliflags/flags.go:
```diff
@@ -152,11 +152,12 @@ with a non-zero status code and further statements are not executed. The
-When not specified, table rows are printed as tab-separated values (TSV).`,
+Selects how to display table rows in results.
+Possible options: tsv, csv, pretty, records, sql, html.
+When not specified, table rows are printed as tab-separated values (tsv).`,
```
It defaults to "pretty" interactively and "tsv" otherwise, right?

bdarnell reviewed (commented):
One thing that postgresql's `\x` and mysql's `\G` have that this PR doesn't is the ability to change the formatting for a single query within a running session. It would be nice to have something similar.

knz committed:
cli/sql: extend the `\set` local command

This patch extends `\set` as follows:

- `\set` can now be used to configure the table display format (`\set
  display_format`).

- `\set` without argument now shows the current value of all options.

knz commented:
> One thing that postgresql's |\x| and mysql's |\G| have that this PR
> doesn't is the ability to change the formatting for a single query
> within a running session. It would be nice to have something similar.

Done. (\set)

>>  		Description: `
> -Causes table rows to be formatted as tables using ASCII art.
> -When not specified, table rows are printed as tab-separated values (TSV).`,
> +Selects how to display table rows in results.
> +Possible options: tsv, csv, pretty, records, sql, html.
> +When not specified, table rows are printed as tab-separated values (tsv).`,
> 
> It defaults to "pretty" interactively and "tsv" otherwise, right?

Thanks for spotting this. Fixed.

mjibson commented:
Review status: 0 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 835 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-KboZn-v824q-LHFwL6v:-KboZn-v824q-LHFwL6w:b-pnmuvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/cli_test.go#L835)):*
> ```Go
> 	// "foo\nbar" | 0
> 	// κόσμε      | 0
> 	// ܈85        | 0
> ```

Can we add a column with a bar in the name to make sure it's escaped correctly?

---

*[pkg/cli/format_table.go, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-Kboaq3zi07dcUTcLnR8:-Kboaq4-laHXPkm6sIdx:b-kx6uuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/format_table.go#L33)):*
> ```Go
> 
> // printQueryOutput takes a list of column names and a list of row contents
> // writes a formatted table to 'w', or "OK" if empty.
> ```

The comment about "OK" if empty seems to be wrong. Should be something about the tag instead?

---

*[pkg/cli/format_table.go, line 106 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-KboeYH2RJsz_W5XJruX:-KboeYH2RJsz_W5XJruY:b-tos26s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/format_table.go#L106)):*
> ```Go
> 						colLabel = ""
> 					}
> 					fmt.Fprintf(w, "%-*s | %s\n", maxColWidth, colLabel, line)
> ```

TIL `%-*s`.  This doesn't appear to escape `|` from colLabel. Is that a problem?

---

*[pkg/cli/sql.go, line 292 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-Kboi6cXhWKQgoZYMkTj:-Kboi6cXhWKQgoZYMkTk:b-lqz74) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/sql.go#L292)):*
> ```Go
> 	}
> 	if err := opt.set(c, args[1:]); err != nil {
> 		return c.invalidOptionChange(errState, args[0])
> ```

Would it be more useful here to display the error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 14 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 835 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-KboZn-v824q-LHFwL6v:-Kbp3lLzdYi2YcKqW936:b-xexdkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/cli_test.go#L835)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Can we add a column with a bar in the name to make sure it's escaped correctly?
</blockquote></details>

It's not escaped. See below. I still added an example here to clarify.

---

*[pkg/cli/format_table.go, line 33 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-Kboaq3zi07dcUTcLnR8:-Kbp4-U1Lze2kJA0F6xW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/format_table.go#L33)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The comment about "OK" if empty seems to be wrong. Should be something about the tag instead?
</blockquote></details>

Done.

---

*[pkg/cli/format_table.go, line 106 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-KboeYH2RJsz_W5XJruX:-Kbp40GNoeF3N5rtfdfw:bnwryb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/format_table.go#L106)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

TIL `%-*s`.  This doesn't appear to escape `|` from colLabel. Is that a problem?
</blockquote></details>

It's what pg does. This format is already inherently lossy for multiple reasons, so I didn't want to spend too much time to fix it.

---

*[pkg/cli/sql.go, line 292 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-Kboi6cXhWKQgoZYMkTj:-Kbp6iGBsILm0z9SrVOt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad19bc211b456a96c874e9d9bb068eb830dc0fa3/pkg/cli/sql.go#L292)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Would it be more useful here to display the error?
</blockquote></details>

Done.

---

*[pkg/cli/cliflags/flags.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-KbR1VGfcgC64N1ak1pY-r1-160:-Kbp6HESqiWIYBgXQKRz:bo6apac) ([raw file](https://github.com/cockroachdb/cockroach/blob/b884d1420dd3015204fbc42d4810d3b62c26475e/pkg/cli/cliflags/flags.go#L160)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It defaults to "pretty" interactively and "tsv" otherwise, right?
</blockquote></details>

Indeed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171)*
<!-- Sent from Reviewable.io -->

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 14 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13171#-:-KbpiXIWRH7yPWsr1eXR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR!
